File: company-name-matching/java-examples/generate-match-report.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.FileReader;
import java.net.HttpURLConnection;
import java.net.URLEncoder;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * Reads company names from a text file, calls Interzoid's
 * getcompanymatchadvanced API to generate SimKeys, and prints
 * clusters of matching SimKeys (two or more records) as:
 *     Input,SimKey
 * with a blank line between clusters.
 *
 * Suggested file name: company-match-report.java
 * Run with:            java CompanyMatchReport
 */
class CompanyMatchReport {

    // Replace this with your API key from https://www.interzoid.com/manage-api-account
    private static final String API_KEY = "YOUR_API_KEY_HERE";

    // Input file containing one company name per line
    private static final String INPUT_FILE_NAME = "sample-input-file.txt";

    // Holds a single input company name and its SimKey
    private static class Record {
        String input;
        String simKey;

        Record(String input, String simKey) {
            this.input = input;
            this.simKey = simKey;
        }
    }

    // Maps the JSON returned by the getcompanymatchadvanced API
    private static class ResponseData {
        String simKey;
        String code;
        String credits;
    }

    public static void main(String[] args) {
        List<Record> records = new ArrayList<>();

        // Read the input file line by line
        try (BufferedReader reader = new BufferedReader(new FileReader(INPUT_FILE_NAME))) {
            String company;

            while ((company = reader.readLine()) != null) {
                // Skip blank lines
                if (company.trim().isEmpty()) {
                    continue;
                }

                // Call API and get SimKey
                ResponseData apiResult;
                try {
                    apiResult = callGetCompanyMatchAdvanced(company);
                } catch (Exception e) {
                    System.out.println("Error calling API for: " + company + " - " + e.getMessage());
                    continue;
                }

                if (apiResult == null || apiResult.simKey == null || apiResult.simKey.isEmpty()) {
                    // If there is no SimKey, skip this record
                    continue;
                }

                records.add(new Record(company, apiResult.simKey));
            }

        } catch (IOException e) {
            System.out.println("Error opening input file: " + e.getMessage());
            return;
        }

        // If there are no records, nothing to do
        if (records.isEmpty()) {
            System.out.println("No records with similarity keys found.");
            return;
        }

        //------------------------------------------------------------------
        // Sort the records strictly by SimKey only, so identical SimKeys
        // become adjacent and we can easily cluster.
        //------------------------------------------------------------------
        Collections.sort(records, Comparator.comparing(r -> r.simKey));

        //------------------------------------------------------------------
        // Walk through the sorted list and collect clusters of records that
        // share the same SimKey. Only print clusters that contain two or more
        // records. Each line is "Input,SimKey", with blank lines between clusters.
        //------------------------------------------------------------------
        String currentKey = null;
        List<Record> cluster = new ArrayList<>();

        for (Record rec : records) {
            if (currentKey == null || !rec.simKey.equals(currentKey)) {
                // SimKey changed: flush previous cluster
                printCluster(cluster);
                cluster.clear();

                currentKey = rec.simKey;
                cluster.add(rec);
            } else {
                // Same SimKey as current cluster, so add to it
                cluster.add(rec);
            }
        }

        // Flush final cluster
        printCluster(cluster);
    }

    /**
     * Helper to print a cluster if it has two or more records.
     * Each record is printed as: Input,SimKey
     * with a blank line between clusters.
     */
    private static void printCluster(List<Record> cluster) {
        if (cluster.size() < 2) {
            return;
        }
        for (Record r : cluster) {
            System.out.printf("%s,%s%n", r.input, r.simKey);
        }
        System.out.println();
    }

    /**
     * Calls the getcompanymatchadvanced API for a given company name.
     * Returns ResponseData with SimKey, Code, Credits.
     */
    private static ResponseData callGetCompanyMatchAdvanced(String company) throws IOException {
        String companyParam = URLEncoder.encode(company, StandardCharsets.UTF_8.toString());

        String apiUrl =
            "https://api.interzoid.com/getcompanymatchadvanced?license=" + API_KEY +
            "&company=" + companyParam +
            "&algorithm=model-v4-wide";

        URL url = new URL(apiUrl);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");

        int status = conn.getResponseCode();
        BufferedReader in = new BufferedReader(
            new InputStreamReader(
                status >= 200 && status < 300
                    ? conn.getInputStream()
                    : conn.getErrorStream(),
                StandardCharsets.UTF_8
            )
        );

        StringBuilder responseBuilder = new StringBuilder();
        String line;
        while ((line = in.readLine()) != null) {
            responseBuilder.append(line);
        }
        in.close();
        conn.disconnect();

        String json = responseBuilder.toString().trim();
        if (json.isEmpty()) {
            return null;
        }

        ResponseData data = new ResponseData();
        data.simKey = extractJsonValue(json, "SimKey");
        data.code = extractJsonValue(json, "Code");
        data.credits = extractJsonValue(json, "Credits");

        return data;
    }

    /**
     * Minimal JSON value extractor for flat JSON like:
     * {"SimKey":"...","Code":"Success","Credits":"1"}
     * This avoids external libraries to keep it parallel to the Go example.
     */
    private static String extractJsonValue(String json, String key) {
        String quotedKey = "\"" + key + "\"";
        int keyIndex = json.indexOf(quotedKey);
        if (keyIndex == -1) return "";

        int colonIndex = json.indexOf(':', keyIndex + quotedKey.length());
        if (colonIndex == -1) return "";

        int valueStart = colonIndex + 1;
        while (valueStart < json.length() && Character.isWhitespace(json.charAt(valueStart))) {
            valueStart++;
        }

        // Handle quoted string value
        if (valueStart < json.length() && json.charAt(valueStart) == '"') {
            valueStart++;
            StringBuilder sb = new StringBuilder();
            boolean escaping = false;

            for (int i = valueStart; i < json.length(); i++) {
                char c = json.charAt(i);
                if (escaping) {
                    sb.append(c);
                    escaping = false;
                } else if (c == '\\') {
                    escaping = true;
                } else if (c == '"') {
                    break;
                } else {
                    sb.append(c);
                }
            }
            return sb.toString();
        } else {
            // Fallback for non-string values (not expected here)
            int valueEnd = valueStart;
            while (valueEnd < json.length()
                    && json.charAt(valueEnd) != ','
                    && json.charAt(valueEnd) != '}'
                    && !Character.isWhitespace(json.charAt(valueEnd))) {
                valueEnd++;
            }
            return json.substring(valueStart, valueEnd).trim();
        }
    }
}

File: individual-name-matching/kotlin-examples/generate-match-report.kt
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

// Replace this with your API key from https://www.interzoid.com/manage-api-account
const val API_KEY = "YOUR_API_KEY_HERE"

// Input file containing one full name per line
const val INPUT_FILE_NAME = "sample-input-file.txt"

// Simple data class to hold the input name and its similarity key (SimKey)
data class Record(
    val input: String,
    val simKey: String
)

fun main() {

    val records = mutableListOf<Record>()

    val inputFile = File(INPUT_FILE_NAME)
    if (!inputFile.exists()) {
        println("Input file not found: $INPUT_FILE_NAME")
        return
    }

    // Read each line as a full name and call the API
    inputFile.forEachLine { line ->
        val fullName = line.trim()

        // Skip blank lines
        if (fullName.isEmpty()) return@forEachLine

        val simKey = getFullNameSimKey(fullName)

        // Skip if no SimKey returned
        if (simKey.isNullOrEmpty()) return@forEachLine

        records.add(Record(input = fullName, simKey = simKey))
    }

    if (records.isEmpty()) {
        println("No records with similarity keys found.")
        return
    }

    //----------------------------------------------------------------------
    // Sort strictly by SimKey so identical keys are adjacent and easy to cluster
    //----------------------------------------------------------------------
    records.sortBy { it.simKey }

    //----------------------------------------------------------------------
    // Walk through the sorted list and build clusters by SimKey.
    // Only print clusters with two or more records.
    // Each line: "Input,SimKey"
    // Clusters separated by blank lines.
    //----------------------------------------------------------------------
    var currentKey: String? = null
    val cluster = mutableListOf<Record>()

    fun printCluster(c: List<Record>) {
        if (c.size < 2) return     // Only print clusters with 2+ records
        for (r in c) {
            println("${r.input},${r.simKey}")
        }
        println()                  // blank line between clusters
    }

    for (rec in records) {
        if (rec.simKey != currentKey) {
            if (cluster.isNotEmpty()) {
                printCluster(cluster)
                cluster.clear()
            }
            currentKey = rec.simKey
        }
        cluster.add(rec)
    }

    // Flush final cluster
    if (cluster.isNotEmpty()) {
        printCluster(cluster)
    }
}

/**
 * Calls Interzoid's getfullnamematch API for a single full name
 * and returns the similarity key (SimKey) as a String.
 * Returns null or empty if there is an error or no SimKey.
 */
fun getFullNameSimKey(fullName: String): String? {
    val encodedName = URLEncoder.encode(fullName, "UTF-8")

    val apiUrl = "https://api.interzoid.com/getfullnamematch" +
            "?license=$API_KEY" +
            "&fullname=$encodedName"

    return try {
        val url = URL(apiUrl)
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"

        val body = connection.inputStream.bufferedReader().use { it.readText() }
        connection.disconnect()

        // For this example, manually extract "SimKey" from the JSON response.
        // Response is expected to look like:
        // {"SimKey":"...","Code":"Success","Credits":"..."}
        extractSimKey(body)
    } catch (e: Exception) {
        println("Error calling API for \"$fullName\": ${e.message}")
        null
    }
}

/**
 * Minimal helper to extract "SimKey" from a simple JSON object string.
 * For production, use a JSON library like Jackson or kotlinx.serialization.
 */
fun extractSimKey(json: String): String? {
    val key = "\"SimKey\""
    val keyIndex = json.indexOf(key)
    if (keyIndex == -1) return null

    val colonIndex = json.indexOf(':', keyIndex)
    if (colonIndex == -1) return null

    val firstQuote = json.indexOf('"', colonIndex + 1)
    if (firstQuote == -1) return null

    val secondQuote = json.indexOf('"', firstQuote + 1)
    if (secondQuote == -1) return null

    return json.substring(firstQuote + 1, secondQuote)
}
